### PR TITLE
feat: Added redis for caching the posts data

### PR DESCRIPTION
**Pull request by krishnaacharyaa**
## Summary

The loading of the posts is taking nearly 5s to get from the server, which is quite huge, so added a redis layer on top of it, so that the posts data is cached initially and when the new posts is added it is invalidated.

## Description

It is aimed to bring down the latency of the fetch call from the backend server. Now it is taking nearly 5sec without redis for all the fetches.

### NOTE !!
You would have to set a redis in your local machine or create a redis account in the [upstash](https://console.upstash.com/redis) and place the generated url in the .env `REDIS_URL`'s value

## Images
**Before Redis**
image

**After Redis**
image

## Prerequisites

- [x] Have you followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by chinmaykunkikar** [comment]
image
The api request URL is bad here

**Comment by krishnaacharyaa** [comment]
> image The api request URL is bad here

Where exactly did you hit this? I am playing around with CORS, and quite a few changes in the endpoints and to whitelist and stuff...
Production app is now down. 

image
This has recently slipped in the production, causing all the issue, in production if I give proper CORS_URL works as expected.
But in preview deployments as every deployment name changes, it causes deadlock, 
Frontend needs VITE_URL from backend
Backend needs this CORS_URL of frontend... basically causing a deadlock. In production it is fine though... because the domain name doesn't change, 

Any take on this? I have removed this CORS_URL change and trying to make it work, production app is now down, not sure how it was running previously :)

**Comment by krishnaacharyaa** [comment]
@chinmaykunkikar @bajajcodes  Need your reviews

**Comment by krishnaacharyaa** [comment]
Succesfully cross checked in local, working just fine as expected and checked with the remote redis cache aswell

File: backend/utils/cache-posts.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { redis } from '../config/redis.js';
+import { REDIS_PREFIX } from './constants.js';
+
+// Helper function to check if Redis is available
+function isRedisEnabled() {
+  return redis !== null;
+}
+
+export async function getKeyFromCache(key) {
```
Comment (bajajcodes) [inline backend/utils/cache-posts.js:9]: Better to use `retrieveDataFromCache`, and `storeDataInCache`

Comment (bajajcodes) [reply backend/utils/cache-posts.js:9]: [OPTIONAL]: You can create a single service which will take care of redis connection, redis client and supporting methods/features.

File: backend/config/redis.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import dotenv from 'dotenv';
```
Comment (bajajcodes) [inline backend/config/redis.js:1]: configs are not used for clients, better to use service for redis client.
To get environment-vars, you can use config.

File: backend/config/redis.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      url: process.env.REDIS_URL,
+      disableOfflineQueue: true,
+    }).connect();
+    console.log('Redis Connected: ' + process.env.REDIS_URL);
+  } catch (error) {
+    console.error('Error connecting to Redis:', error.message);
+    redis = null;
+  }
+} else {
+  console.log('Redis not configured, cache disabled.');
```
Comment (bajajcodes) [inline backend/config/redis.js:18]: If env is missing throw error or return null.

File: backend/utils/cache-posts.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+export async function setKeyInCache(key, data) {
+  if (!isRedisEnabled()) return; // Skip cache if Redis is not available
+
+  const cacheKey = `${REDIS_PREFIX}:${key}`;
+  await redis.set(cacheKey, JSON.stringify(data));
+}
+
+export async function invalidateKeyInCache(key) {
```
Comment (bajajcodes) [inline backend/utils/cache-posts.js:27]: Alternative name: deleteKeyFromCache

**Review by bajajcodes — CHANGES_REQUESTED** [review]
Better put redis under a services, and use single client, and connect to redis at the start of the server.

**Comment by krishnaacharyaa** [comment]
All the comments are addressed, Ek aur baar review kardeejiye please :)

**Pull request merged**